cosmos_rl/utils/parallelism_map.py

```diff
@@ -1258,32 +1332,37 @@ async def generate_parallelized_shard_send_insts_for_policy(
                             insts_for_param_name.append(
                                 WeightSyncInstruction(
                                     p_rank, r, p_tensor_split_strategys
-                                )
+                                ).__dict__
                             )
                         for p in r_assignments:
                             if (
                                 dest_name
-                                not in self.rollout_from_policy_insts_meta[r_rank]
+                                not in self.rollout_from_policy_insts_meta[p_rank][
+                                    r_rank
+                                ]
                             ):
-                                self.rollout_from_policy_insts_meta[r_rank][
+                                self.rollout_from_policy_insts_meta[p_rank][r_rank][
                                     dest_name
-                                ] = []
-                            self.rollout_from_policy_insts_meta[r_rank][
+                                ] = ParallelizedShardMapper.multiprocessing_manager.list()
+
+                            self.rollout_from_policy_insts_meta[p_rank][r_rank][
                                 dest_name
                             ].append(
                                 WeightSyncInstruction(
                                     p, r_rank, r_tensor_split_strategys
-                                )
+                                ).__dict__
                             )
                 if insts_for_param_name:
                     insts_for_group.append(
-                        WeightSyncInstructionsPerParam(dest_name, insts_for_param_name)
+                        WeightSyncInstructionsPerParam(
+                            dest_name, insts_for_param_name
+                        ).__dict__
                     )
             if insts_for_group:
                 policy_to_rollout_insts.append(
-                    WeightSyncInstructionsGroup(insts_for_group)
+                    WeightSyncInstructionsGroup(insts_for_group).__dict__
                 )
-        return policy_to_rollout_insts
+        return msgpack.packb(policy_to_rollout_insts)
```

Comment:
should return the original python obj here, and do the `msgpack.pack` encoding later in the http server (right before returning the responses)

Reply:
good point. but actually I'm not sure if it will be slow, since python uses pickle/unpickle to transfer between multiprocessing, native python types (like Dict, List, str) might be fast enough. maybe need to benchmark

Reply:
Yeah, we can test it then.